Add Save Macro Image into SaveModule #230

nanli-emory opened the pull request:
add 'saveMarco' function to save marco image in saveModule
1. add saveAssociatedImage function to save the associated image by key
2. add saveMacro function to save the 'marco' image from slide's associated images

nanli-emory committed:
Save Macro Image module #230
1. add saveAssociatedImage function to save the associated image by key
2. add saveMacro function to save the 'marco' image from slide's associated images

choosehappy commented on histoqc/SaveModule.py:
```diff
@@ -36,6 +36,39 @@ def saveFinalMask(s, params):
 
 
+def saveAssociatedImage(s, key:str, dim:int):
+    logging.info(f"{s['filename']} - \tsaveMacro")
```
is this line strictly correct? this function would seem to be able to save any type of associated image. it is being called from savemacro, thus the actual image pulled is dependent upon the 'key' provided

are there other images that we will want to save, or is it only the macro which is available? if only the macro this may make more sense as a single function

i think there are two options here:
1) merge these two functions
2) saveMacro in this info needs to be - 'save{key}'? or something of the link?

nanli-emory replied:
I checked some slides' associated_images. There are  'label', 'thumbnail', and 'macro' images. Let's keep **saveAssociatedImage** function just in case we need to read other associated images.

choosehappy replied:
interesting........okay good idea. maybe we can even justify adding 'label' as a save option?

choosehappy commented on histoqc/SaveModule.py:
```diff
@@ -36,6 +36,39 @@ def saveFinalMask(s, params):
+    
+    # get asscociated image by key
+    associated_img = osh.associated_images[key]
+    (width, height)  = associated_img.size
```
this is okay since i don't think we have a standard for histoqc, but when you look at the other tools i think you're more likely to see "nrow" and "ncol" instead of "width" and "height". just an FYI

nanli-emory replied:
should **ncol** correspond to **height** and **nrow** correspond to **width**, right?

choosehappy replied:
i think its nrow = height, ncol = width?

choosehappy commented on histoqc/config/config_light.ini:
```diff
@@ -18,6 +18,7 @@ steps= BasicModule.getBasicStats
 ;    BrightContrastModule.getBrightnessByChannelinColorSpace:YUV
 ;    DeconvolutionModule.separateStains
     SaveModule.saveFinalMask
+    SaveModule.saveMacro
```
great! 

2 more questions:
1) should we add it into the other config files (in particular think 2.1 is the new default which people should use, and use that as a template)

2) is this going to be visible/selectable from the UI? i think it just involves adding the "_macro.png" suffix somewhere in JS?

nanli-emory replied:
I will add **SaveModule.saveFinalMask** into config.ini and add **_macro.png** suffix in JS.

nanli-emory committed:
1. add "SaveModule.saveMacro" config into other config.ini files
2. add "_marco.png" into  'global_config.js' files to show up the marco images.
3. correct the info message "saveMacro" to "save{key.capitalize()}"
